src/vs/base/browser/ui/tree/stickyScroll.ts

```diff
@@ -0,0 +1,459 @@
+	readonly node: ITreeNode<T, TFilterData>;
+	readonly startIndex: number;
+	readonly endIndex: number;
+	readonly height: number;
+	readonly position: number;
+}
+
+class StickyScrollState<T, TFilterData, TRef> extends Disposable {
+
+	constructor(
+		readonly stickyNodes: StickyScrollNode<T, TFilterData>[] = []
+	) {
+		super();
+	}
+
+	get count(): number { return this.stickyNodes.length; }
+
+	equal(state: StickyScrollState<T, TFilterData, TRef>): boolean {
+		return this.count === state.count &&
+			equals(this.stickyNodes, state.stickyNodes, (r1, r2) => r1.position === r2.position) &&
+			equals(this.stickyNodes, state.stickyNodes, (r1, r2) => r1.node.element === r2.node.element) &&
+			equals(this.stickyNodes, state.stickyNodes, (r1, r2) => r1.startIndex === r2.startIndex) &&
+			equals(this.stickyNodes, state.stickyNodes, (r1, r2) => r1.height === r2.height) &&
+			equals(this.stickyNodes, state.stickyNodes, (r1, r2) => r1.endIndex === r2.endIndex);
+	}
+
+	addDisposable(disposable: IDisposable): void {
+		this._register(disposable);
+	}
+}
+/* hot-reload:patch-prototype-methods */
+export class StickyScrollController<T, TFilterData, TRef> extends Disposable {
+
+	private maxNumberOfStickyElements: number;
+	private readonly maxWidgetViewRatio = 0.4;
+
+	private readonly _widget: StickyScrollWidget<T, TFilterData, TRef>;
+	private readonly _treeHelper: TreeModelHelpers<T, TFilterData, TRef>;
+
+	private get firstVisibleNode() {
+		const index = this.view.firstVisibleIndex;
+
+		if (index < 0 || index >= this.view.length) {
+			return undefined;
+		}
+
+		return this.view.element(index);
+	}
+
+	constructor(
+		private readonly tree: AbstractTree<T, TFilterData, TRef>,
+		model: ITreeModel<T, TFilterData, TRef>,
+		private readonly view: List<ITreeNode<T, TFilterData>>,
+		renderers: TreeRenderer<T, TFilterData, TRef, any>[],
+		private readonly treeDelegate: IListVirtualDelegate<ITreeNode<T, TFilterData>>,
+		options: IAbstractTreeOptions<T, TFilterData> = {},
+	) {
+		super();
+
+		const stickyScrollOptions = this.validateStickySettings(options);
+		this.maxNumberOfStickyElements = stickyScrollOptions.stickyScrollMaxItemCount;
+
+		this._treeHelper = new TreeModelHelpers(model);
+		this._widget = this._register(new StickyScrollWidget(view.getScrollableElement(), view, model, renderers, treeDelegate));
+
+		this._register(view.onDidScroll(() => this.setState()));
+		this._register(view.onDidChangeContentHeight(() => this.setState()));
+		this._register(tree.onDidChangeCollapseState(() => this.setState()));
+		this._register(tree.onDidChangeSelection(() => this.setState()));
+		this._register(tree.onDidChangeFocus(() => this.setState()));
+
+		this.setState();
+	}
+
+	private setState() {
+		const firstVisibleNode = this.firstVisibleNode;
+
+		// Don't render anything if there are no elements
+		if (!firstVisibleNode || this.tree.scrollTop === 0) {
+			this._widget.setState(new StickyScrollState());
```

Comment:
`new StickyScrollState()` will still allocate an object and an array, at a minimum. Could we use `undefined` instead to represent the "no state" option?